code/stippler.py

```diff
@@ -0,0 +1,232 @@
+    bbox = np.array([xmin, xmax, ymin, ymax])
+    ratio = (xmax-xmin)/(ymax-ymin)
+
+    # Interactive display
+    if args.interactive:
+
+        # Setup figure
+        fig = plt.figure(figsize=(args.figsize, args.figsize/ratio),
+                         facecolor="white")
+        ax = fig.add_axes([0, 0, 1, 1], frameon=False)
+        ax.set_xlim([xmin, xmax])
+        ax.set_xticks([])
+        ax.set_ylim([ymin, ymax])
+        ax.set_yticks([])
+        scatter = ax.scatter(points[:, 0], points[:, 1], s=1, color=".25")
+
+        def update(frame):
+            global points
+            regions, points = voronoi.centroids(points, density)
+            Pi = points.astype(int)
+            X = np.maximum(np.minimum(Pi[:, 0], density.shape[1]-1), 0)
+            Y = np.maximum(np.minimum(Pi[:, 1], density.shape[0]-1), 0)
+            sizes = (args.pointsize[0] +
+                     (args.pointsize[1]-args.pointsize[0])*density[Y, X])
+            scatter.set_offsets(points)
+            scatter.set_sizes(sizes)
+            bar.update()
+
+            # Save stipple points and stippled image
+            if (frame == args.n_iter-1 and
+                    not os.path.exists(dat_filename) or args.save):
+                np.save(dat_filename, points)
+                plt.savefig(pdf_filename)
+                plt.savefig(png_filename)
+
+        bar = tqdm.tqdm(total=args.n_iter)
+        animation = FuncAnimation(fig, update,
+                                  repeat=False, frames=args.n_iter-1)
+        plt.show()
+
+    elif not os.path.exists(dat_filename) or args.force:
+        for i in tqdm.trange(args.n_iter):
+            regions, points = stipple(points, density)
+
+    if (args.save or args.display) and not args.interactive:
+        fig = plt.figure(figsize=(args.figsize, args.figsize/ratio),
+                         facecolor="white")
+        ax = fig.add_axes([0, 0, 1, 1], frameon=False)
+        ax.set_xlim([xmin, xmax])
+        ax.set_xticks([])
+        ax.set_ylim([ymin, ymax])
+        ax.set_yticks([])
+        scatter = ax.scatter(points[:, 0], points[:, 1], s=1, color=".25")
+        Pi = points.astype(int)
+        X = np.maximum(np.minimum(Pi[:, 0], density.shape[1]-1), 0)
+        Y = np.maximum(np.minimum(Pi[:, 1], density.shape[0]-1), 0)
+        sizes = (args.pointsize[0] +
+                 (args.pointsize[1]-args.pointsize[0])*density[Y, X])
+        scatter.set_offsets(points)
+        scatter.set_sizes(sizes)
+
+        # Save stipple points and tippled image
+        if not os.path.exists(dat_filename) or args.save:
+            np.save(dat_filename, points)
+            plt.savefig(pdf_filename)
+            plt.savefig(png_filename)
+
+        if args.display:
+            plt.show()
+
+    # Plot voronoi regions
+    # for region in vor.filtered_regions:
+    #     vertices = vor.vertices[region, :]
+    #     ax.plot(vertices[:, 0], vertices[:, 1], linewidth=.5, color='.5' )
```

Comment:
This serves at showing how you can plot Voronoi cells if you want. I would prefer to keep it as it is.